File: lib/java/src/main/java/uapi/internal/generation/GenerateRandomArray.java
package uapi.internal.generation;

import java.util.ArrayList;
import java.util.List;

import uapi.internal.types.UTypeDeclaration;

public class GenerateRandomArray {
    public static Object generateRandomArray(Object blueprintValue, boolean useBlueprintValue,
            List<UTypeDeclaration> typeParameters, GenerateContext ctx) {
        final var nestedTypeDeclaration = typeParameters.get(0);

        if (useBlueprintValue) {
            final var startingArray = (List<Object>) blueprintValue;

            final var array = new ArrayList<Object>();
            for (final var startingArrayValue : startingArray) {
                final var value = nestedTypeDeclaration
                        .generateRandomValue(startingArrayValue, true, ctx);

                array.add(value);
            }

            return array;
        } else {
            final var length = ctx.randomGenerator.nextCollectionLength();

            final var array = new ArrayList<Object>();
            for (int i = 0; i < length; i += 1) {
                final var value = nestedTypeDeclaration.generateRandomValue(null, false, ctx);

                array.add(value);
            }

            return array;
        }
    }
}

File: lib/py/uapi/internal/generation/GenerateRandomArray.py
from typing import TYPE_CHECKING, cast


if TYPE_CHECKING:
    from uapi.RandomGenerator import RandomGenerator
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration
    from uapi.internal.generation.GenerateContext import GenerateContext


def generate_random_array(ctx: 'GenerateContext') -> list[object]:
    nested_type_declaration = ctx.type_parameters[0]

    if ctx.use_blueprint_value:
        starting_array = cast(list[object], ctx.blueprint_value)

        array = []
        for starting_array_value in starting_array:
            value = nested_type_declaration.generate_random_value(
                ctx.copy(blueprint_value=starting_array_value))

            array.append(value)

        return array
    else:
        length = ctx.random_generator.next_collection_length()

        array = []
        for i in range(length):
            value = nested_type_declaration.generate_random_value(ctx)

            array.append(value)

        return array
